### PR TITLE
Feature/crowd 124 mail on comments

### PR DESCRIPTION
**Pull request by wThomas84**
Project creator retrieves a mail when somebody commented her project; Comments by herself won't trigger a mail.

### CODE REVIEW

File: crowdsource-core/src/main/java/de/asideas/crowdsource/controller/CommentController.java [hunk 1]
```diff
@@ -30,40 +23,18 @@
 public class CommentController {
 
     @Autowired
-    private ProjectRepository projectRepository;
+    private CommentService commentService;
```
Comment (CanardSauvage) [inline crowdsource-core/src/main/java/de/asideas/crowdsource/controller/CommentController.java]: I don't like names ending with *Service. The class name does not indicate what the class does. This also a well known Anti-Pattern, like "Bean", "Impl" etc.

File: crowdsource-core/src/main/java/de/asideas/crowdsource/domain/model/CommentEntity.java [hunk 2]
```diff
@@ -31,10 +31,9 @@
     @LastModifiedDate
     private DateTime lastModifiedDate;
 
-    public CommentEntity(ProjectEntity projectEntity, UserEntity userEntity, String comment) {
-
+    public CommentEntity(ProjectEntity projectEntity, UserEntity commentingUser, String comment) {
         this.project = projectEntity;
-        this.user = userEntity;
+        this.user = commentingUser;
```
Comment (CanardSauvage) [inline crowdsource-core/src/main/java/de/asideas/crowdsource/domain/model/CommentEntity.java]: I like this as it make it more clear which user this is.

File: crowdsource-core/src/main/java/de/asideas/crowdsource/service/CommentService.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+@Service
+public class CommentService {
+
+    private CommentRepository commentRepository;
+    private ProjectService projectService;
+    private UserService userService;
+    private UserNotificationService userNotificationService;
+
+    @Autowired
+    public CommentService(CommentRepository commentRepository, ProjectService projectService, UserService userService, UserNotificationService userNotificationService) {
```
Comment (CanardSauvage) [inline crowdsource-core/src/main/java/de/asideas/crowdsource/service/CommentService.java]: This is not the project style - change it everywhere or don't use constructor injection.

File: crowdsource-core/src/main/java/de/asideas/crowdsource/service/ProjectService.java [hunk 4]
```diff
@@ -160,7 +160,7 @@ void pledgeProjectUsingPostRoundBudget(ProjectEntity projectEntity, UserEntity u
         LOG.debug("Project pledged using post round budget: {}", pledgeResult);
     }
 
-    private ProjectEntity loadProjectEntity(String projectId) {
+    protected ProjectEntity loadProjectEntity(String projectId) {
```
Comment (CanardSauvage) [inline crowdsource-core/src/main/java/de/asideas/crowdsource/service/ProjectService.java]: I would prefer package private and not protected.

File: crowdsource-core/src/test/java/de/asideas/crowdsource/service/CommentServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.List;
+
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.reset;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class CommentServiceTest {
```
Comment (CanardSauvage) [inline crowdsource-core/src/test/java/de/asideas/crowdsource/service/CommentServiceTest.java]: * The names of the test-method should reflect what they are testing. Eg. addComment_shouldCallSaveAndNotify 
* The setup of the test should be in the test-method, not in the Before/Init method. I was wondering "why does the commentService find Projects for a project? The answer: Because it was mocked before.
* You should extract private methods to make the setup more readable eg. givenUserServiceFindsAUserForExistingMail or at least mockUserService()

File: crowdsource-example/src/test/java/de/asideas/crowdsource/service/MailTest.java [hunk 6]
```diff
@@ -200,6 +202,36 @@ public void notifyCreatorAndAdminOnProjectModification() {
 
     }
 
+    @Test
+    public void notifyCreatorOnComment_mailShouldContainAbridgedCommentAndItsCreatorName() {
```
Comment (CanardSauvage) [inline crowdsource-example/src/test/java/de/asideas/crowdsource/service/MailTest.java]: This is not a readable test. The setup could be extracted to make the test more readable.

File: crowdsource-example/src/test/java/de/asideas/crowdsource/service/MailTest.java [hunk 7]
```diff
@@ -200,6 +202,36 @@ public void notifyCreatorAndAdminOnProjectModification() {
...
+                "Zu Deinem Projekt:\n\n%s\n\n" +
+                "Mit freundlichen Grüßen\n" +
+                "Dein CrowdSource Team";
+
+        userNotificationService.notifyCreatorOnComment(comment);
+
+        final List<SimpleMailMessage> capturedMessages = getMessagesFromMailSender();
+        assertThat(capturedMessages.size(), is(1));
+
+        assertThat(capturedMessages.get(0).getText(), is(String.format(expMessage,
```
Comment (CanardSauvage) [inline crowdsource-example/src/test/java/de/asideas/crowdsource/service/MailTest.java]: eg. this call is really hard to read. Better would be thenXyz or assertTextContainsSomething

**Comment by CanardSauvage** [comment]
The overall quality is ok, you could improve some minor things, like the readability of the tests.

One other thing: The package structure of the project could be improved. First, it as a technical structure. Second, it does not have a homogenous deepth of the packages. And third, there are packages with only one Class. Forth, the "Controller" belong to the "presentation", so you have presentation-stuff in your domain-model.

File: crowdsource-core/src/main/java/de/asideas/crowdsource/controller/CommentController.java (hunk 1 above)
Comment (wThomas84) [inline crowdsource-core/src/main/java/de/asideas/crowdsource/controller/CommentController.java]: This is an infrastructure level class that represents an entry point into the application level of the project, containing methods for each use case concerning comments. It handles interactions with  persistence modules and delegates execution to concerning domain entities and services (which in this case is rather trivial). It shall not be confused with a domain service, implementing specific logic.
An application service is intended to be used by different clients to provide the use cases a client might like to consume. Currently there is just one client - a RESTful controller; another client could be an artifact from within the application itself, or another controller providing comments of a project as a file or an email sink within the application, that allows to add comments by mail answers - you get the idea.
So, until now, I have no better name than "~service". I'd also like "CommentApplicationService" but it's too long. What about CommentAS? It seems too short.. open for ideas..

File: crowdsource-core/src/main/java/de/asideas/crowdsource/service/CommentService.java (hunk 3 above)
Comment (wThomas84) [inline crowdsource-core/src/main/java/de/asideas/crowdsource/service/CommentService.java]: Look around, it has already been changed in lots of classes. My reasons here are:
 * Easier testability with mockito (as it doesn't try to resolve transitive @Autowired dependencies)
 * Dependencies of the class enforced clearer, as it won't be possible to instantiate it without them or at least to stumble upon them

File: crowdsource-core/src/main/java/de/asideas/crowdsource/controller/CommentController.java (hunk 1 above)
Comment (CanardSauvage) [inline crowdsource-core/src/main/java/de/asideas/crowdsource/controller/CommentController.java]: As you may know a class should do a single thing. You can leave the logic there for now, or you could keep the logic in the controller. Better names can be achived if you split the class into two - a *Service is always a dump for stuff!

* LoadCommentsByProject could be directy part of Project eg. project.getComments() and if this does not work because of using a Document-Store as Database you could just implement this in the repository, eg. commentRepository.findByProject
* You could have a CommentSearch-Class for complex queries and a CommentSave-Class for saving comments, this is a well known pattern for CRUD-Applications.

File: crowdsource-core/src/main/java/de/asideas/crowdsource/service/CommentService.java (hunk 3 above)
Comment (CanardSauvage) [inline crowdsource-core/src/main/java/de/asideas/crowdsource/service/CommentService.java]: I'm not talking about wether you should one or another style, I don't mind.

I just opened two classes and both did NOT use const. injection, so this is no homogenous style in the application.

**Comment by wThomas84** [comment]
"Forth, the "Controller" belong to the "presentation", so you have presentation-stuff in your domain-model." Please elaborate.

There exists a `presentation package`, containing the presentation model. It could be moved up in its level to exist next to the core domain model.

File: crowdsource-core/src/main/java/de/asideas/crowdsource/controller/CommentController.java (hunk 1 above)
Comment (wThomas84) [inline crowdsource-core/src/main/java/de/asideas/crowdsource/controller/CommentController.java]: "*Service is always a dump for stuff!"
I disagree, as it has a specific purpose; Structure! And: "homogenous style in the application" on an architectual level. This is worth much more as which kind of Spring's injection mechanism is utilized. It gives the overview of use cases available in that domain. It segregades parts of different responsibilities within the application and thus allows finally to concentrate the actual and often so hard to maintain logic of the application to be isolated in one place, allowing it to be tested independently from any infrastructure or client concerns. 
This is one reason I like the technical structure first and eventually sub divide by business concerns. Having this structure in mind it is easy to directly navigate to core business artifacts when you want to understand business stuff or navigate directly to - lets say messaging stuff -when you wanna know how a messaging middleware is connected. Its kind of a house where you can always feel like home, due to you know where to find all the things you are interested in.

Comment (CanardSauvage) [inline crowdsource-core/src/main/java/de/asideas/crowdsource/controller/CommentController.java]: We can discuss this in person if you really want. 

But I can't believe that you really argue about classes called *Service to be not stupid.

Also I can't believe that a fan of DDD is using technical packages/namespaces. Both will become a huge issue in bigger projects with more classes. In small projects it does not matter that much.

And your argumentation for *Services is really bad - come on, really? This violates so many principles of clean code. AND as is told you - I don't care which kind of injection mechanism you use.

> "It gives the overview of use cases available in that domain"
Wrong, this is a dump. A class should have a single purpose and not do "everything in the domain". 

> "maintain logic of the application to be isolated in one place"
Again wrong. If this is your argument, you can skip the service and have the logic in the controller. There it would be in one place.

> "it is easy to directly navigate to core business artifacts when you want to understand business stuff or navigate directly to" 
This only applies for small projects. If you change you application, you will not want to jump between packages as you normalle change one thing in the business logic.
Also in bigger applications you will to many classes of one kind eg. controller or "service" and this won't help you to maintain any kind of maintainability.


TL;DR: You can merge this, as my notes were meant to help you improve the structure of your application.

**Comment by CanardSauvage** [comment]
Presentation: Everything concerning the UI or the presentation to the outside, eg. Controller or pure Frontend. I can show you AFK what I mean.

**Comment by CanardSauvage** [comment]
TL;DR: You can merge this, as my notes were meant to help you improve the structure of your application.
